### ISSUE TITLE
N + 1 round trip problem

### ISSUE THREAD

**Issue by itaied246** [graphql-python/graphene-sqlalchemy#35]
Does this library handle nested models (joins) in a single query from the server to the DB?
 For example

user {
  id
  posts {
    id
  }
}

**Comment by yfilali** [comment]
No, it is something you'll need to handle on your own. See a few code samples to deal with this here: https://github.com/graphql-python/graphene/issues/348

**Comment by yfilali** [comment]
In case you are still looking for a complete solution:
https://yacine.org/2017/02/27/graphqlgraphene-sqlalchemy-and-the-n1-problem/

**Comment by Cito** [comment]
Thank you for that blog post @yfilali, that was really helpful.

Your `optimize_resolve` method works pretty well for me. By using `joinedload` instead of `subqueryload`, I was able to reduce a deeply nested GraphQL query into a single database query. I had to add all paths, however, not only the leaves (simply removed the `if not p.isLeaf` part when assembling the joins).

Another small problem is that in `resolve_related` you need to translate the field names from GraphQL (camel case) to SQLAlchemy (snake case) with `field = to_snake_case(field)`, otherwise it doesn't work with table columns with underscores in them. You can import the `to_snake_case` method from `graphene.utils.str_converters`.

I feel something like this should be really integrated into graphene-sqlalchemy. Or at least it should provide some helper functions which make it easy to implement such optimized querying, with various loading strategies as options. After all, one of the main advantages of GraphQL over REST is that it solves the N+1 problem. This advantage would be greater if it existed also on the level of the database, not only on the network level.

**Comment by chanind** [comment]
Another option is to use a lazy-loader for SQLAlchemy that is more intelligent about bulk loading. We wrote a custom loader that will inspect the DB session and bulk-load all relations on similar models whenever a relation is lazy-loaded. The result should be the equivalent to using `subqueryload` on all relations upfront, but without needing any custom code to do so. The loader we wrote is available here in case it's helpful to anyone else: https://github.com/operator/sqlalchemy_bulk_lazy_loader

**Comment by flewellyn** [comment]
@yfilali Is that code you wrote for optimize_resolve under the same MIT license as the rest of the library?  I would love to make use of it.

**Comment by yfilali** [comment]
@flewellyn absolutely!

**Comment by mekhami** [comment]
@yfilali I can't seem to access your website, is it still up? Would you mind posting the contents here if not?

**Comment by yfilali** [comment]
@mekhami it seems up at the moment so it could have been temporary or regional (I didn't get any alerts either)

**Comment by mekhami** [comment]
@yfilali sorry to bother again but it's still down and I'm not able to access it. Could you post the text of it here or somewhere? I remember it being extremely valuable but didn't have the time then to implement it.

**Comment by ocavue** [comment]
@mekhami I think it's valuable and I can't access it too. I used google cache to read it several months ago, but the catch itself is 404 now, maybe because the origin page is unavailable for too long. Fortunately, I saved the cache: [GraphQL, Graphene, SqlAlchemy and the N+1 problem _ Yacine's copy.htm.txt](https://github.com/graphql-python/graphene-sqlalchemy/files/2500047/GraphQL.Graphene.SqlAlchemy.and.the.N%2B1.problem._.Yacine.s.copy.htm.txt) (github don't allow me to upload a .htm file so I add .txt at the end)

@yfilali I will delete this comment and the cached attachment if you are not ok with this.

**Comment by ocavue** [comment]
@mekhami I think it's valuable and I can't access it too. I read it from google cache several months ago and I do think it's worth to read. But the catch itself if 404 now, maybe because the origin page is down for too long. I saved the cache just in case: 
[GraphQL_Yacine.htm.txt](https://github.com/graphql-python/graphene-sqlalchemy/files/2500076/GraphQL_Yacine.htm.txt) (github don't allow .htm file so I add .txt at the end)

@yfilali I will delete the cache page if you are not ok with this.

**Comment by ocavue** [comment]
@mekhami I think it's valuable and I can't access it too. I read it from google cache several months ago and I do think it's worth to read. But the catch itself if 404 now, maybe because the origin page is down for too long. I saved the cache just in case: [GraphQL_Yacine.zip](https://github.com/graphql-python/graphene-sqlalchemy/files/2500080/GraphQL_Yacine.zip)
 (github don't allow .htm file so I compress it)

@yfilali I will delete the cache page if you are not ok with this.

**Comment by yfilali** [comment]
@mekhami it's completely fine. Sorry about that. I'll have to find time to work on my personal site's uptime at some point :)

**Comment by iksteen** [comment]
A bit late to the party. Also: shameless self-promotion. But I've just published an article about this subject which includes a solution where you specify what to eagerly load in a node's metadata on a per-field basis: https://blurringexistence.net/drafts/graphene-sqlalchemy-n-plus-1.html

**Comment by flewellyn** [comment]
@iksteen Your eager loader looks fascinating.  Question: will it be able to handle custom-defined object types that are not SQLAlchemy models?  I have a couple defined to handle special cases, like geoalchemy2 geometries.

**Comment by jnak** [comment]
Just a heads that there is a PR currently opened that addresses the N+1 problem. Would love to get your input. Cheers.

See https://github.com/graphql-python/graphene-sqlalchemy/pull/253#pullrequestreview-309278130

**Comment by flewellyn** [comment]
Intriguing.  When that's available, I will have to compare it with @yfilali's solution for performance and usability.

**Comment by sreerajkksd** [comment]
Did we solve this in https://github.com/graphql-python/graphene-sqlalchemy/pull/253 and https://github.com/graphql-python/graphene-sqlalchemy/pull/254 ?

**Comment by jnak** [comment]
Yes that's available as of [`2.3.0`](https://github.com/graphql-python/graphene-sqlalchemy/releases/tag/2.3.0). See release notes how to enable it. Would love feedback on performance and usability.

**Comment by sreerajkksd** [comment]
I tested the batching option and it works flawlessly. 

I have one more related question:

From the given example, If `user` is meant to return multiple users, then it still continue to iterate on all these users and then there posts to return the data. Can that be solved in anyway ?

To be more clear: can we have selectinload for users than just posts ?

**Comment by adrianschneider94** [comment]
Hey, I tested #253, however it's targeting Relay, right? At least I couldn't make it work without it.
I guess I reimplemented something that was discussed here before :D
But it works and maybe could be of use for somebody.

https://gist.github.com/adrianschneider94/90f662ffab9dce06e2f291579ad480b7

Usage:

```python
class Query(graphene.ObjectType):
    get_all = graphene.List(ModelSchema)

    def resolve_get_all(self, info):
        query: Query = ModelSchema.get_query(info)
        query = smart_load(query, info)
        return query.all()
```
You can select the loading strategy:
```python
query = smart_load(query, info, strategy="select-in")
# or
query = smart_load(query, info, strategy="lazy")
# or 
query = smart_load(query, info, strategy="joined")
```

Best wishes!

**Comment by adrianschneider94** [comment]
And here the dataloader approach which is superior I guess.


Concept:
  * `SmartSQLAlchemyObjectType` inherits from `SQLAlchemyObjectType` and creates a `promise.DataLoader` for each relationship
  * The data loaders are created with a factory function that only depends on the SQLAlchemy attribute
  * All data loaders are stored in a class variable as mapping `RelationshipProperty` -> data loader
  * Additionally `SmartSQLAlchemyObjectType` overwrites the default resolver for each schema class which uses the data loaders.

I get the session from a ContextVar but I guess there should be a solution where you use the session of `root` in `sqlalchemy_default_resolver`.

I haven't written tests yet, but it works really well so far.

```python
from graphene.types.resolver import default_resolver
from graphene_sqlalchemy import SQLAlchemyObjectType
from promise import Promise
from promise.dataloader import DataLoader
from sqlalchemy import inspect, tuple_
from sqlalchemy.orm import RelationshipProperty
from sqlalchemy.orm.attributes import InstrumentedAttribute

from my_api.context import use_session


def get_identity(obj):
    return inspect(obj).identity


def sqlalchemy_default_resolver(attname, default_value, root, info, **kwargs):
    parent_type = info.parent_type

    if hasattr(root, "_sa_class_manager"):
        if attname in root._sa_class_manager:
            attribute = root._sa_class_manager[attname]
            property = attribute.property
            if property in parent_type.graphene_type.data_loaders.keys():
                data_loader = parent_type.graphene_type.data_loaders[property]
                return data_loader.load(get_identity(root))

    return default_resolver(attname, default_value, root, info, **kwargs)


def create_relationship_loader(attribute: InstrumentedAttribute):
    class RelationshipLoader(DataLoader):
        def batch_load_fn(self, keys):
            def resolver(resolve, reject):
                session = use_session()
                Remote = attribute.entity.class_
                Local = attribute.parent.class_
                primary_keys = tuple(key.expression for key in inspect(Local).primary_key)

                result = session \
                    .query(tuple_(*primary_keys), Remote) \
                    .join(attribute) \
                    .filter(tuple_(*primary_keys).in_(keys)) \
                    .order_by(*primary_keys) \
                    .all()

                res = [[v for k, v in result if k == key or k == key[0]] for key in keys]
                return resolve(res)

            return Promise(resolver)

    return RelationshipLoader


class SmartSQLAlchemyObjectType(SQLAlchemyObjectType):
    data_loaders = {}

    class Meta:
        abstract = True

    @classmethod
    def __init_subclass_with_meta__(cls, *args, **kwargs):
        model = kwargs['model']
        for key, attribute in model._sa_class_manager.local_attrs.items():
            property = attribute.property
            if isinstance(property, RelationshipProperty):
                data_loader = create_relationship_loader(attribute)(cache=False, max_batch_size=1000)
                cls.data_loaders[property] = data_loader

        if not "default_resolver" in kwargs:
            kwargs['default_resolver'] = sqlalchemy_default_resolver

        super().__init_subclass_with_meta__(*args, **kwargs)
```

**Comment by thaonc97** [comment]
Could anyone provide a working example with batching please? I 've tried to figure it out for days but still no luck.

**Comment by thaonc97** [comment]
OK I figured it out, the `batching` option doesn't work with the current release of `SQLAlchemy` (1.4.20), I reinstalled `SQLAlchemy `to version 1.4.20 and its work fine. Haven't tested on others.

**Comment by PaulSchweizer** [comment]
I am also a bit confused about the batching feature. 
I inspected the tests and I can see **two SELECT statements** where I would **expect only one**.  

https://github.com/graphql-python/graphene-sqlalchemy/blob/master/graphene_sqlalchemy/tests/test_batching.py#L121

Printing the logging messages that are captured in the test, I can see two SELECT statements:

```python
[
    'BEGIN (implicit)', 
    'SELECT articles.id AS articles_id, articles.headline AS articles_headline, articles.pub_date AS articles_pub_date, articles.reporter_id AS articles_reporter_id \nFROM articles', 
    '[generated in 0.00017s] ()', 
    'SELECT reporters.id AS reporters_id, (SELECT CAST(count(reporters.id) AS INTEGER) AS count_1 \nFROM reporters) AS anon_1, reporters.first_name AS reporters_first_name, reporters.last_name AS reporters_last_name, reporters.email AS reporters_email, reporters.favorite_pet_kind AS reporters_favorite_pet_kind \nFROM reporters \nWHERE reporters.id IN (?, ?)', 
    '[generated in 0.00059s] (1, 2)'
]
```

I get the same result with SQLAlchemy==1.3.18 and 1.4.35 and also ran sqltap to verify this behavior.

We have solved the n+1 issue similar to how @adrianschneider94 demonstrated above so I thought that batching could make our custom solution superfluous but that seems to not be the case.

Am I maybe misunderstanding what batching is supposed to do?

**Comment by PaulSchweizer** [comment]
So I read up on it a bit more and it makes sense to me now:

Batching uses the "select in" loading mechanic of sqlalchemy which loads all related objects upfront by using the primary keys returned from the first query. It then emits a second SELECT statement to load the related objects in bulk as opposed to loading each one of them individually at the point of access. This is just the gist however and not the whole truth, better to read up the official docs: https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#selectin-eager-loading

So in conclusion if you want just ONE SINGLE SELECT statement, you have to implement your own solution like mentioned above.

**Comment by flipbit03** [comment]
I've activated `batching = True` on all my graphene-sqlalchemy translated models, and whenever I try to access some relationships, the graphql resolver is failing with "future is attached to a different loop" 

This is the GraphQL Query:

```graphql
{
readConstruct(constructId:"ad500676-5503-aac0-ec38-d3097958c6d0")
   {
    id
    constructParts { <--- this is a relationship
      id
      part { <-- this is a relationship
        id
      }
    }
  }
}
```

I get an error on the `part` resolver (but also in several other places in the system, basically it's not working correctly in most places)
```
Task <Task pending name='Task-347' coro=<ExecutionContext.resolve_field.<locals>.await_result() running at /home/cadu/.local/share/virtualenvs/kernel-backend-N3YlRYFF/lib/python3.10/site-packages/graphql/execution/execute.py:625> cb=[gather.<locals>._done_callback() at /home/cadu/.pyenv/versions/3.10.6/lib/python3.10/asyncio/tasks.py:720]> got Future <Future pending> attached to a different loop
```
